Improve experience for creating signers

piratekev opened the pull request:
This PR adds a new package `farcaster-js-cli` that is included in the `farcaster-js` package for creating signers. It also updates the neynar package to simplify signer creation. These changes are reflected in the top level README so developers can easily find this information.

Example CLI runs:
```
$ farcaster-js create-signer
There are 2 options for creating a signer.

Developer Managed:
- A signer is generated by interacting with the KeyGateway Contract on Optimism (0x00000000fc56947c7e7183f8ca4b62398caadf0b).
- You must have a small amount of funds (~0.001) in your account's custody address on Optimism.
- You will need to provide the mnemonic for your account.
- This signer can be used anywhere (except with Neynar).

Neynar Managed:
- Neynar creates a signer using your developer mnemonic.
- Approval is done through the Warpcast App by any user.
- Payment is required in the app using Warps or a $0.99 in-app purchase.
- A Neynar API Key is necessary.
- This signer is exclusive to Neynar APIs.

Which type of signer will you be creating?
 (Use arrow keys)
❯ developer
  neynar
? Enter your mnemonic: your mnemonic goes here
Transaction written to OP Mainnet. Check txn status at https://optimistic.etherscan.io/tx/<txn hash>
Transaction Confirmed! Your signer is ready to use.
```

```
$ farcaster-js create-signer
There are 2 options for creating a signer.

Developer Managed:
- A signer is generated by interacting with the KeyGateway Contract on Optimism (0x00000000fc56947c7e7183f8ca4b62398caadf0b).
- You must have a small amount of funds (~0.001) in your account's custody address on Optimism.
- You will need to provide the mnemonic for your account
- This signer can be used anywhere (except with Neynar).

Neynar Managed:
- Neynar creates a signer using your developer mnemonic.
- Approval is done through the Warpcast App by any user.
- Payment is required in the app using Warps or a $0.99 in-app purchase.
- A Neynar API Key is necessary.
- This signer is exclusive to Neynar APIs.

Which type of signer will you be creating?
 (Use arrow keys)
  developer
❯ neynar
? Enter your Neynar API Key: <neynar api key>
? Enter the mnemonic for the developer: your mnemonic goes here
Creating a new Neynar signer...
 ▄▄▄▄▄▄▄ ▄▄▄ ▄▄ ▄▄ ▄  ▄ ▄  ▄▄▄▄▄▄▄
 █ ▄▄▄ █  ▄██ ▀█ █▄█▀█▄▀▄▄ █ ▄▄▄ █
 █ ███ █ ██ ▄▀▄ ▀█▀ ▄   █▄ █ ███ █
 █▄▄▄▄▄█ █ █ ▄ █▀▄ █ █▀▄ █ █▄▄▄▄▄█
 ▄   ▄ ▄▄█▀ ▄█  ▄ █▀▀▀█▀  ▄▄▄▄▄  ▄
 █▀▀▀▀▄▄▄▀█▀█▀▄▄█▀▀█ ▄▀▄█ ▀   █▀█
 ███ █▄▄▀ ▀  ▀█ █▄  █ ███ ▀█ ▄▄  ▄
  ▄▄██▀▄▀█▀ ▀▄▀█▄ ▀█▀█▀▀█▄█▄█ ▀▄▄
 ▄█▄ ▀█▄▀█▄█  ▄ ▀▀█ ▄ ██▀█▀█▀█▀ ▄
  ▀█ ▄█▄▄▀▀▄  ▄██▄▄▀ ▄ ▄█ ▀ ▄ █▀▄
 ▄▀▀▄▄ ▄ ▄ ▀██▀ ▀ ▀█▀█▀ ▀█▀▄█▄▄ ▄▄
 ▀▀▄▄█▀▄▄▀█  █▀▀ █▄▄█  ▄█ ▄ ▀▄▀▄█
 ▄▄▄██▄▄ █▄▄█ ▀█▄ █ ▀██▄▀█████  ▀▀
 ▄▄▄▄▄▄▄ ██▀ ▀▀▀██▀█ █▀ ▀█ ▄ █▄▀▄
 █ ▄▄▄ █ ▄ ▀▄█▀█ █   ▀ █ █▄▄▄█▄ █
 █ ███ █  ▀▄▀▄█ ▄ ▄▀▀█▀ ██ █████▀
 █▄▄▄▄▄█ ▄█▄▀▀█▀▀██▀▄▄█▄▄█       ▄
▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀

URL: https://client.warpcast.com/deeplinks/signed-key-request?token=<register token>
Signer UUID: <signer uuid>

Scan the QR code above or visit the URL on a logged in device to approve signer

-----------------------------------

Once approved, you can start using your signer with Neynar to write data to Farcaster
( ●    ) Approved! Now you can start using this signer.
Signer UUID: <signer uuid>
✔
```


It also refactors signer creation in the neynar package, see below:

Creating a neynar signer before:
```
  // create signer
  const signer = await client.v2.createSigner();

  // create signature
  const signature = await generateSignature(
    signer.public_key,
    signerFid,
    privateKey,
    deadline,
  );

  // register signer
  const registeredSigner = await client.v2.registerSigner(
    signer.signer_uuid,
    userBotFid,
    deadline,
    signature,
  );
```

Creating a signer neynar after:
```
const registeredSigner = await client.v2.createAndRegisterSigner(
  signerFid,
  deadline,
  privateKey,
);
```


Open questions:

1. Should we mask input for mnemonics?
2. Should we warn users about protecting signer private keys / signer uuids?

piratekev committed:
chore(farcaster-js-neynar): fix generate signer script

piratekev committed:
feat(farcaster-js-neynar): improve user experience for creating a neynar signer

affects: @standard-crypto/farcaster-js-neynar

piratekev committed:
docs(farcaster-js-neynar): improve signer documentation

piratekev committed:
docs(farcaster-js-neynar): improve docs

gsgalloway commented on packages/farcaster-js-neynar/src/v2/NeynarV2APIClient.ts:
```diff
@@ -135,18 +127,28 @@ export class NeynarV2APIClient {
-    signerUuid: string,
+  public async createAndRegisterSigner(
     fid: number,
     deadline: number,
```
Do you think it would be helpful to set a default value for `deadline`, or just make it more confusing?

piratekev replied:
Yeah that's a good idea, what do you think the default should be? I'm torn because I'd like it to be essentially infinite, but neynar doesn't provide a way to revoke the access for it.

piratekev replied:
30 days seemed like a good compromise. I made it clear in the examples as well.

gsgalloway reviewed (approved):
Nice

piratekev committed:
feat(farcaster-js-neynar): set default deadline for signer

affects: @standard-crypto/farcaster-js-neynar

piratekev committed:
feat(farcaster-js): add cli interface to create signers

affects: @standard-crypto/farcaster-js-cli, @standard-crypto/farcaster-js-neynar,
@standard-crypto/farcaster-js

add cli to create signers (developer and neynar-managed)
add new package farcaster-js-cli
update
signer flow & examples in farcaster-js-neynar
pull in latest api spec for farcaster-js-neynar

piratekev committed:
chore(farcaster-js-cli): update package structure

affects: @standard-crypto/farcaster-js-cli, @standard-crypto/farcaster-js

piratekev committed:
chore(farcaster-js-cli): docs & cleanup

affects: @standard-crypto/farcaster-js-cli, @standard-crypto/farcaster-js

piratekev committed:
chore(farcaster-js-neynar): cleanup

affects: @standard-crypto/farcaster-js-neynar

piratekev committed:
chore(farcaster-js): docs & cleanup

affects: @standard-crypto/farcaster-js-cli, @standard-crypto/farcaster-js

piratekev committed:
fix(farcaster-js-neynar): cleanup

affects: @standard-crypto/farcaster-js-neynar

piratekev committed:
chore(farcaster-js): update lockfile

piratekev commented:
@gsgalloway ready for review

gsgalloway commented on examples/neynarSigner.ts:
```diff
@@ -0,0 +1,22 @@
+
+const developerMnemonic = 'your farcaster recovery phrase';
+// default deadline is 30 days - set longer if needed
+// const deadline = Math.floor(Date.now() / 1000) + 30 * 86400;
```
```suggestion
```
I'd imagine that the docstring for the `createSigner` method should be good enough for someone who needs to change the deadline, wdyt?

gsgalloway commented on examples/neynarSigner.ts:
```diff
@@ -0,0 +1,22 @@
+// create signer
+const signer = await client.v2.createSigner(
+  developerMnemonic,
+//   deadline,
```
```suggestion
```

gsgalloway commented on packages/farcaster-js-cli/README.md:
````diff
@@ -0,0 +1,103 @@
+? Enter your mnemonic: your mnemonic goes here
+Transaction written to OP Mainnet. Check txn status at https://optimistic.etherscan.io/tx/<txn hash>
+Transaction Confirmed! Your signer is ready to use.
+```
````
It feels a bit odd to me that the README for this package would include two full transcripts of interacting with the CLI.

I think one of the benefits of making the CLI so high touch as you did is that it reduces the need for so much documentation.

My impression is that all the reader of the README needs to know is that there are two paths for creating a signer, either using Neynar or not, and that the CLI will walk them through both.

piratekev replied:
agreed, removed.

gsgalloway commented on packages/farcaster-js-cli/package.json:
```diff
@@ -0,0 +1,72 @@
+{
+  "name": "@standard-crypto/farcaster-js-cli",
+  "version": "6.0.3",
```
Copypasta

gsgalloway commented on packages/farcaster-js-cli/package.json:
```diff
@@ -0,0 +1,72 @@
+{
+  "name": "@standard-crypto/farcaster-js-cli",
+  "version": "6.0.3",
+  "description": "A collection of tools for interacting with the Farcaster social network.",
```
Copypasta

gsgalloway commented on packages/farcaster-js-cli/package.json:
```diff
@@ -0,0 +1,72 @@
+  },
+  "homepage": "https://github.com/standard-crypto/farcaster-js-cli",
+  "bugs": "https://github.com/standard-crypto/farcaster-js/issues",
+  "author": "Gavi Galloway",
```
```suggestion
  "author": "Kevin Ayling",
```

gsgalloway commented on packages/farcaster-js-cli/tsconfig.build.json:
```diff
@@ -0,0 +1,11 @@
+    "include": [
+        "./src/**/*.ts"
+    ]
+}
```
Looks like this file is equivalent to `./tsconfig.json`

piratekev replied:
removed

gsgalloway commented on packages/farcaster-js-cli/package.json:
```diff
@@ -0,0 +1,72 @@
+    "postpack": "rm README.md",
+    "publish": "yarn npm publish",
+    "pre-tshy": "mv tsconfig.json tsconfig.json.bak && mv tsconfig.build.json tsconfig.json",
+    "post-tshy": "mv tsconfig.json tsconfig.build.json && mv tsconfig.json.bak tsconfig.json"
```
I suspect we don't need these in this case

gsgalloway commented on packages/farcaster-js-cli/package.json:
```diff
@@ -0,0 +1,72 @@
+    "clean": "yarn clean:dist",
+    "clean:dist": "rm -rf dist",
+    "cz": "yarn run -T cz",
+    "prepack": "cp ../../README.md .",
+    "postpack": "rm README.md",
```
Curious why we're copying the README down from the project root into this package

piratekev replied:
copypasta

gsgalloway commented on packages/farcaster-js-cli/package.json:
```diff
@@ -0,0 +1,72 @@
+    "dist/**"
+  ],
+  "scripts": {
+    "build": "yarn run pre-tshy && yarn run -T tshy && yarn run post-tshy",
```
Here's a dumb line of questioning. But I'm going to ask it anyway

- I know we need a commonjs build of this package for node to be able to run it
- What do we need the ESM build of this package for?
- If it turns out we only need the commonjs version, then do we need to use `tshy`, or will regular `tsc` suffice?

piratekev replied:
This worked - until the farcaster-js package tried to import it using ESM. So I think we still need both.

gsgalloway replied:
Ah, makes sense

gsgalloway commented on packages/farcaster-js-cli/src/bin.ts:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env node
+/* eslint-disable eslint-comments/disable-enable-pair */
+/* eslint-disable no-console */
```
```suggestion
```

gsgalloway commented on README.md:
```diff
@@ -52,6 +52,15 @@ npm install axios @standard-crypto/farcaster-js
 
+***Signers:***
+
+This package include a CLI for creating signers. You can run the code below to generate a signer:
```
```suggestion
This package includes a CLI for creating signers. You can run the code below to generate a signer:
```

gsgalloway commented on README.md:
````diff
@@ -52,6 +52,15 @@ npm install axios @standard-crypto/farcaster-js
+```
+
+See examples of running the CLI in [farcaster-js-cli/](./packages/farcaster-js-cli/README.md)
+
````
This is the first touch point for a dev learning about signers. Maybe also add something like this?
```suggestion
Additionally, signers may be created programmatically without use of the CLI -- see the examples below.
```

gsgalloway commented on packages/farcaster-js-cli/src/cli.ts:
```diff
@@ -0,0 +1,87 @@
+    });
+
+  program.parse(process.argv);
+}
```
Very slick!

gsgalloway commented on packages/farcaster-js-neynar/docs/classes/v2.CastApi.md:
```diff
@@ -101,9 +101,9 @@ BaseAPI.configuration
 ### cast
 
-▸ **cast**(`requestParameters`, `options?`): `Promise`<`AxiosResponse`<[`CastResponse`](../interfaces/v2.CastResponse.md), `any`\>\>
+▸ **cast**(`requestParameters`, `options?`): `Promise`<`AxiosResponse`<`CastResponse`, `any`\>\>
```
Links broken in docs

piratekev replied:
ugh, will look into this.

gsgalloway commented on packages/farcaster-js-neynar/src/v2/openapi/spec.yaml:
```diff
@@ -145,6 +145,10 @@ components:
         - verifications
         - activeStatus
       properties:
+        object:
+          type: string
+          enum:
+            - user
```
`User.object == 'user'`

Wild. Any clue why they added this?

piratekev replied:
No idea. There are some quirks to the specs they generate.

gsgalloway commented on packages/farcaster-js-neynar/src/v2/openapi/spec.yaml:
```diff
@@ -270,7 +286,6 @@ components:
       type: object
       required:
         - signer_uuid
-        - text
```
Huh. `text` is now optional when posting a cast?

piratekev replied:
🤷‍♂️

gsgalloway commented on packages/farcaster-js-neynar/src/v2/openapi/spec.yaml:
```diff
@@ -1027,3 +1335,101 @@ paths:
           $ref: "#/components/responses/403Response"
         "500":
           $ref: "#/components/responses/500Response"
+  /farcaster/user/custody-address:
```
We should probably add issues to track creating bindings for these new API endpoints

piratekev replied:
agreed. Updates have been a bit more challenging than I would like them to be.

gsgalloway replied:
Would you be up for making a quick github issue to track the new endpoints?

gsgalloway commented on packages/farcaster-js/src/cli.ts:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env node
+import { runCli } from '@standard-crypto/farcaster-js-cli';
+void runCli();
```
Slick!

gsgalloway commented on README.md:
````diff
@@ -69,6 +78,34 @@ const cast = await client.v2.publishCast(signerUuid, 'This is a test cast.');
 ```
 <!-- AUTO-GENERATED-CONTENT:END -->
+***Creating a Neynar Signer:***
+<!-- AUTO-GENERATED-CONTENT:START (CODE:src=./examples/neynarSigner.ts) -->
````
It's odd that this example exists twice -- once here in the `farcaster-js` project root and then again in `farcaster-js-neynar`. Should we delete the duplicate and just have this README reference the one from the neynar package? Should we include this example in the top-level README at all, versus pointing users down to the neynar README if they need to create a neynar signer?

piratekev replied:
I think pointing to the neynar readme makes sense - there's info on the CLI for quickly getting a signer in the top level readme. Anyone wanting to create neynar signers programmatically will be checking the docs for neynar anyways.

gsgalloway commented on packages/farcaster-js-neynar/examples/createSigner.ts:
```diff
@@ -1,22 +1,22 @@
-const privateKey = 'mnemonic for signer';
-const deadline = Math.floor(Date.now() / 1000) + 86400; // one day from now
+const developerMnemonic = 'your farcaster recovery phrase';
+// default deadline is 30 days - set longer if needed
+// const deadline = Math.floor(Date.now() / 1000) + 30 * 86400;
```
```suggestion
```

gsgalloway commented on packages/farcaster-js-neynar/examples/createSigner.ts:
```diff
@@ -1,22 +1,22 @@
+// create signer
+const signer = await client.v2.createSigner(
+  developerMnemonic,
+//   deadline,
```
```suggestion
```

gsgalloway commented on packages/farcaster-js-neynar/package.json:
```diff
@@ -57,6 +59,7 @@
     "@types/chai": "^4.3.9",
     "@types/chai-as-promised": "^7.1.5",
     "@types/prompts": "^2.4.7",
+    "@types/qrcode": "^1",
```
Still used?

piratekev replied:
removed

piratekev replied:
it's used in the example

gsgalloway replied:
Ah yes, mb

gsgalloway commented on packages/farcaster-js-neynar/package.json:
```diff
@@ -65,6 +68,7 @@
     "mocha": "^10.2.0",
     "openapi-types": "^12.1.3",
     "openapi-typescript": "^6.7.0",
+    "qrcode": "^1.5.3",
```
Still used?

piratekev replied:
removed

piratekev replied:
it's used in the example

gsgalloway commented on packages/farcaster-js-neynar/package.json:
```diff
@@ -47,6 +47,8 @@
   "author": "Kevin Ayling",
   "license": "MIT",
   "dependencies": {
+    "commander": "^11.1.0",
+    "ethers": "^6.8.1",
```
```suggestion
```

gsgalloway commented on packages/farcaster-js-neynar/scripts/generateSigner.ts:
```diff
@@ -1,54 +1,33 @@
 /* eslint-disable no-console */
 /* eslint-disable @typescript-eslint/restrict-template-expressions */
-import { NeynarAPIClient, generateSignature } from '../src/index.js';
+import { NeynarAPIClient, waitForNeynarSignerApproval } from '../src/index.js';
```
Do we still need this script now that the CLI exists?

piratekev replied:
nope, between the cli and examples this is covered

gsgalloway commented on packages/farcaster-js-neynar/package.json:
```diff
@@ -90,7 +94,7 @@
     "test": "yarn test:integration",
     "test:integration": "mocha",
-    "signer": "env TS_NODE_COMPILER_OPTIONS='{\"module\": \"commonjs\" }' ts-node -r dotenv/config scripts/generateSigner.ts",
+    "signer": "ts-node-esm -r dotenv/config scripts/generateSigner.ts",
```
Still needed?

piratekev replied:
removed

gsgalloway reviewed (commented):
Great work!

piratekev committed:
chore(farcaster-js-cli): review comments & cleanup

affects: @standard-crypto/farcaster-js-cli, @standard-crypto/farcaster-js-neynar,
@standard-crypto/farcaster-js

piratekev committed:
chore(farcaster-js): fix deps for examples

affects: @standard-crypto/farcaster-js-neynar

piratekev commented:
@gsgalloway ready for another review. Looking into what went wrong with docs generation now.

piratekev committed:
docs(farcaster-js-neynar): attempt to fix broken links in docs

affects: @standard-crypto/farcaster-js-neynar

gsgalloway commented on packages/farcaster-js-neynar/src/v2/NeynarV2APIClient.ts:
```diff
@@ -315,12 +321,12 @@ export class NeynarV2APIClient {
   public async reactToCast(
     signerUuid: string,
-    reaction: ReactionType,
+    reaction: 'like' | 'recast',
```
Noting that this would necessitate a major version bump.

Is this just to reduce the friction from developers having to import ReactionType from the `/v2` path?

piratekev replied:
Yeah that's right

gsgalloway reviewed (approved):
Great! Happy to merge once docs generation is in a good spot, and depending on whether you want a major bump for the ReactionType.like change or not

piratekev committed:
chore(farcaster-js): merge conflicts

affects: @standard-crypto/farcaster-js-hub-rest, @standard-crypto/farcaster-js-neynar,
@standard-crypto/farcaster-js

piratekev committed:
chore(farcaster-js): bump package.json versions

affects: @standard-crypto/farcaster-js-hub-rest, @standard-crypto/farcaster-js-neynar,
@standard-crypto/farcaster-js

piratekev committed:
chore(farcaster-js-hub-rest): skip failing tests

affects: @standard-crypto/farcaster-js-hub-rest
